bin/setup.sh

```diff
@@ -1,4 +1,30 @@
 #!/bin/sh
 set -e
 
-for PLUG in %PLUGS%; do sudo snap connect %SNAP%:${PLUG}; done
+snap_connect() {
+  if snap connections | grep --quiet "[[:space:]]:$2"; then
```

Comment:
I don't think this works:

```plain
$ /snap/mir-kiosk-neverputt/current/bin/setup.sh 
Warning: Failed to connect 'wayland'. Please connect manually, available providers are:

Warning: Failed to connect 'hardware-observe'. Please connect manually, available providers are:
  - snapd
```

Reply:
It's not obvious what the context is for that. (What providers exist etc.) Could you run:
```
sh -x /snap/mir-kiosk-neverputt/current/bin/setup.sh 
```

Reply:
```plain
$ sh -x /snap/mir-kiosk-neverputt/current/bin/setup.sh 
+ set -e
+ grep --quiet mir-kiosk-neverputt:opengl
+ snap connections
+ grep --quiet mir-kiosk-neverputt:wayland
+ snap connections
+ snap_connect mir-kiosk-neverputt wayland
+ grep --quiet [[:space:]]:wayland
+ snap connections
+ sed -e 1,/slots:/d
+ snap interface wayland
+ available_providers=
+ [ wayland = wayland ]
+ grep --quiet \- ubuntu-frame
+ echo 
+ grep --quiet \- mir-kiosk
+ echo 
+ echo Warning: Failed to connect 'wayland'. Please connect manually, available providers are:\n
Warning: Failed to connect 'wayland'. Please connect manually, available providers are:

+ grep --quiet mir-kiosk-neverputt:hardware-observe
+ snap connections
+ snap_connect mir-kiosk-neverputt hardware-observe
+ grep --quiet [[:space:]]:hardware-observe
+ snap connections
+ sed -e 1,/slots:/d
+ snap interface hardware-observe
+ available_providers=  - snapd
+ [ wayland = hardware-observe ]
+ echo Warning: Failed to connect 'hardware-observe'. Please connect manually, available providers are:\n  - snapd
Warning: Failed to connect 'hardware-observe'. Please connect manually, available providers are:
  - snapd
```

I think you need to look in `snap interfaces` for system-provided slots?

Reply:
The problem here is that `snap connections` doesn't list unconnected slots (and I wasn't testing on fresh install).

I could check `snap interfaces`, but that is supposedly obsolete.

Reply:
Could you just try and `snap connect $1:$2` - and only do something different when that fails?